xapi-ingest-microservice/api.js

```diff
@@ -0,0 +1,245 @@
+ *
+ * @param  {String}           id                    The uuid by which to retrieve the learning activity statement
+ * @param  {Function}         callback              Standard callback function
+ * @param  {Object}           callback.err          An error that occurred, if any
+ * @param  {Statement}        callback.statement    The requested learning activity statement
+ */
+var getStatement = module.exports.getStatement = function(id, callback) {
+  // Parameter validation
+  var validationSchema = joi.object().keys({
+    'id': joi.string().required()
+  });
+
+  var validationResult = joi.validate({
+    'id': id
+  }, validationSchema);
+
+  if (validationResult.error) {
+    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
+  }
+
+  DB.Statement.findById(id).complete(function(err, statement) {
+    if (err) {
+      console.log({'err': err, 'id': id}, 'An error occurred when getting a learning activity statement');
+      return callback({'code': 500, 'msg': err.message});
+    } else if (!statement) {
+      return callback({'code': 404, 'msg': 'Could not find a learning activity statement'});
+    }
+
+    return callback(null, statement);
+  });
+};
+
+/**
+ * Validate and save a learning activity statement
+ *
+ * @param  {Object}           ctx                       Context containing write credentials
+ * @param  {Object}           statement                 The learning activity statement to save
+ * @param  {Function}         callback                  Standard callback function
+ * @param  {Object}           callback.err              An error that occurred, if any
+ */
+var saveStatement = module.exports.saveStatement = function(ctx, statement, callback) {
+  if (!ctx || !ctx.auth) {
+    console.log({'err': err}, 'Prevented storing a learning activity without authentication');
+    return callback({'code': 500, 'msg': 'Prevented storing a learning activity without authentication'});
+  }
+
+  // Verify that no other learning activity statement with the same uuid already exists
+  if (statement.id) {
+    getStatement(statement.id, function(err, retrievedStatement) {
+      if (retrievedStatement) {
+        console.log({'id': statement.id}, 'Attempted to save a learning activity statement that already exists');
+        return callback({'code': 400, 'msg': 'Attempted to save a learning activity statement that already exists'});
+      } else if (err && err.code !== 404) {
+        console.log({'id': statement.id}, 'Unable to verify if learning activity statement already exists');
+        return callback(err);
+      }
+
+      // Validate the statement
+      xapiValidator.validate(statement, function(err) {
+        if (err) {
+          console.log({'err': err}, 'Invalid learning activity statement');
+          return callback({'code': 500, 'msg': err});
+        }
+
+        // When no uuid has been included, generate one
+        statement.id = statement.id || uuid.v4();
```

Comment:
@sandeepmjay given `if (statement.id)` at line 82, the line above will never use `uuid.v4()`